File: service/pyShobuService/src/logic/shobu/moveValidation.py

import logic.shobu.board as B
import logic.shobu.move as M
import logic.shobu.helpers as h
import logic.shobu.token as t

# i think its possible that validatin is done.
# except for validating that the vector is a valid direction.

def validateFullMove( board, fullMove ):
    active = M.moveGetActive( fullMove )
    passive = M.moveGetPassive( fullMove )
    return validatePassiveMove( board, passive ) and validateActiveMove( board, active, M.getSide( passive ), M.getVector( passive ) )

def validatePassiveMove( board, move ):
    side = M.getSide( move )
    spot = M.getSpot( move )
    vector = M.getVector( move )

    if not spotIsInBoard( spot ):
        return False
    subboard = B.getPassiveMoveSubboard( board, side )
    targetToken = B.subboardGetToken( subboard, spot )
    if targetToken != B.boardGetPlayer( board ):
        return False
    unit, steps = h.vectorToUnitAndSteps( vector )
    for i in range( steps ):
        spot = h.addSpotVec( spot, unit )
        if not spotIsInBoard( spot ) or not B.spotIsEmpty( subboard, spot ): 
            return False
    return True

def validateActiveMove( board, move, passiveSide, passiveVector ):
    side = M.getSide( move )
    spot = M.getSpot( move )
    vector = M.getVector( move )

    if not h.compareVec( vector, passiveVector ):
        return False 
    if not spotIsInBoard( spot ):
        return False
    player = B.boardGetPlayer( board )
    subboard = B.getActiveMoveSubboard( board, side, passiveSide )
    targetToken = B.subboardGetToken( subboard, spot )
    if targetToken != player:
        return False
    unit,steps = h.vectorToUnitAndSteps( vector )
    for i in range( steps ):
        spot = h.addSpotVec( spot, unit )
        if not spotIsInBoard( spot ):
            return False
        if not B.spotIsEmpty( subboard, spot ):
            pushTarget = B.subboardGetToken( subboard, spot )
            if not h.tokenIsEnemy( pushTarget, player ) or not validateBeingPushedMovement( subboard, spot, unit, steps - i ):
                return False
    return True

def validateBeingPushedMovement( subboard, spot, unit, steps ):
    for i in range( steps ):
        spot = h.addSpotVec( spot, unit )
        if not spotIsInBoard( spot ):
            return True
        if not B.spotIsEmpty( subboard, spot ):
            return False
    return True

def spotIsInBoard( spot ):
    x,y = spot 
    return x >= 0 and y >= 0 and x < 4 and y < 4

File: client/shobu_client/src/gameLogic/moveValidation.js

import { addSpotVec, compareVec, spotIsInBoard, vectorToUnitAndSteps } from "./spot";
import { getPassiveMoveSubboard, getActiveMoveSubboard, subboardGetToken, tokenBelongsToPlayer, spotIsEmpty } from "./board";
import { tokenIsEnemy } from "./token";


export function validateFullMove( board, fullMove ){
    return validatePassiveMove( board, fullMove.passive ) && 
        validateActiveMove( board, fullMove.active, 
            fullMove.passive.side, fullMove.passive.vector );
}

export function validatePassiveMove( board, move ){
    if( ! spotIsInBoard( move.spot ) ){
        return false;
    }
    let subboard = getPassiveMoveSubboard( board, move.side );
    let targetToken = subboardGetToken( subboard, move.spot );
    if( ! tokenBelongsToPlayer( targetToken, board ) ){
        return false;
    }
    let [ unit, steps ] = vectorToUnitAndSteps( move.vector );
    let spot = move.spot;
    for( let i =0;i< steps;i++){
        spot = addSpotVec( spot, unit );
        if( ! spotIsInBoard( spot ) || ! spotIsEmpty( subboard, spot ) ){
            return false;
        }
    }
    return true;
}

export function validateActiveMove( board, move, passiveSide, passiveVector ){
    if( ! compareVec( move.vector, passiveVector ) || ! spotIsInBoard( move.spot ) ){
        return false;
    }
    let subboard = getActiveMoveSubboard( board, move.side, passiveSide );
    let targetToken = subboardGetToken( subboard, move.spot );
    if( ! tokenBelongsToPlayer( board, targetToken ) ){
        return false;
    }
    let [ unit, steps ] = vectorToUnitAndSteps( move.vector );
    let spot = move.spot;
    for( let i =0;i<steps;i++ ){
        spot = addSpotVec( spot, unit );
        if( ! spotIsInBoard( spot ) ){
            return false;
        }
        if( ! spotIsEmpty( subboard, spot ) ){
            let pushTarget = subboardGetToken( subboard, spot );
            if( ! tokenIsEnemy( pushTarget, board.playerTurn ) || 
                ! validateBeingPushedMovement( subboard, spot, unit, steps ) ){
                return false;
            }
        }
    }
    return true;
}

function validateBeingPushedMovement( subboard, spot, unit, steps ){
    for( let i=0;i< steps;i++){
        spot = addSpotVec( spot, unit );
        if( ! spotIsInBoard( spot ) ){
            return true;
        }
        if( ! spotIsEmpty( subboard, spot ) ){
            return false;
        }
    }
    return true;
}